File: ml/tests/matlab/logistic_regression/test4_circle.m
% Mihail Dunaev
% May 2017
%
% Simple test that tries to separate 2 sets of 2D data on a plane
%   using polynomial (2nd degree) logistic regression. Data is in 
%   'logistic_circle.in'.

function test4_circle()
    addpath('core/');
    [xs, ys] = read_data('../../input/logistic_circle.in');

    syms t;
    w0 = [1; -1; 1; -2; 0.5];
    fun = @(t)(my_cost_with_grad(t, xs, ys));
    w = train(w0, fun);
    
    xpred = [3.5 4; 4 4.1; 2 6];
    ypred = my_predict(xpred, w);
    plot_all(xs, ys, w, xpred, ypred);
    %plot_cost();
end

function [cost, grad] = my_cost_with_grad(t, xs, ys)
    % cost & grad for circle/ellipse logistic regression
    xs0 = xs(ys == 0,:);
    xs1 = xs(ys == 1,:);
    n0 = size(xs0, 1);
    n1 = size(xs1, 1);
    N = n0 + n1;
    cost = (sum(-log(1 ./ (1 + exp(-[xs1 .^ 2 xs1 ones(n1,1)] * t)))) + ...
        sum(-log(1 - (1 ./ (1 + exp(-[xs0 .^ 2 xs0 ones(n0,1)] * t)))))) / N;
    
    if nargout > 1
        num_params = size(t, 1);
        grad = zeros(num_params, 1);
        X = [xs .^ 2 xs ones(N, 1)];
        for i = 1:num_params
            grad(i) = sum(((1 ./ (1 + exp(-[xs .^ 2 xs ones(N,1)] * t))) ...
                - ys) .*  X(:,i)) / N;
        end
    end
end

function ypred = my_predict(xpred, w)
    % function that predicts values for xpred
    % using a circle/ellipse logistic regresssion model with params w
    N = size(xpred, 1);
    % if [x 1] * w >= 0 => class 1, 0 otherwise
    ypred = [xpred .^2 xpred ones(N, 1)] * w >= 0;
end

function plot_all(xs, ys, w, xpred, ypred)
    xs0 = xs(ys == 0,:);
    xs1 = xs(ys == 1,:);
    
    % plot initial data
    figure;
    hold on;
    plot(xs0(:,1), xs0(:,2), 'bo');
    plot(xs1(:,1), xs1(:,2), 'rx');
    
    % set range for x/y axis
    xmin = min(xs(:,1));
    xmax = max(xs(:,1));
    ymin = min(xs(:,2));
    ymax = max(xs(:,2));
    xlim([xmin-1 xmax+1]);
    ylim([ymin-1 ymax+1]);
    
    % plot best fit (3d)
    delta = 0.1;
    [wxs,wys] = meshgrid(xmin:delta:xmax, ymin:delta:ymax);
    syms t;
    sigmoid = @(t)(1 / (1 + exp(-t)));
    wzs = arrayfun(sigmoid, (wxs .^ 2) .* w(1) + (wys .^ 2) .* w(2) + ...
        wxs .* w(3) + wys .* w(4) + w(5));    
    surf(wxs, wys, wzs);
    
    % plot predicted points
    %plot3(xpred(:,1), xpred(:,2), ypred, 'go');
    plot(xpred(:,1), xpred(:,2), 'go');
end

File: ml/tests/python/scikit_learn/logistic_regression/test4_circle.py
"""
Mihail Dunaev
May 2017

Simple test that tries to separate 2 sets of 2D data on a plane
  using polynomial (2nd degree) logistic regression. Data is in 
  'logistic_circle.in'.

For logistic regression I used scikit-learn:
  sklearn.linear_model.LogisticRegression
"""

import matplotlib.pyplot as plt
import numpy as np
from sklearn import linear_model
from scipy.special import expit
from mpl_toolkits.mplot3d import Axes3D

from common.aux import read_data

def test4_circle():
    (xs, ys) = read_data("../../../input/logistic_circle.in")

    # train logistic regression model with sklearn
    lr = linear_model.LogisticRegression(C=1e5)
    lr.fit(np.concatenate((xs ** 2, xs), axis=1), ys)

    # predict some values
    xpred = np.array([3.5, 4, 4, 4.1, 2, 6]).reshape(3, 2);
    ypred = lr.predict(np.concatenate((xpred ** 2, xpred), axis=1))
    print(ypred.tolist())

    plot_all(xs, ys, lr, xpred, ypred, plot3d=True)

def plot_all(xs, ys, lr, xpred, ypred, plot3d=False):
    # switch backend - TODO fix this
    plt.switch_backend("TKagg")
    if plot3d:
        fig = plt.figure(1)
        ax = fig.gca(projection="3d")

    N = len(xs)
    xs0 = np.array([xs[i] for i in xrange(N) if ys[i] == 0])
    xs1 = np.array([xs[i] for i in xrange(N) if ys[i] == 1])

    # plot initial data
    plt.plot(xs0[:,0], xs0[:,1], "bo")
    plt.plot(xs1[:,0], xs1[:,1], "rx")

    # set range for x/y axis
    xmin = xs[:,0].min()
    xmax = xs[:,0].max()
    ymin = xs[:,1].min()
    ymax = xs[:,1].max()
    plt.xlim([xmin - 1, xmax + 1])
    plt.ylim([ymin - 1, ymax + 1])

    # plot best fit (3D)
    if plot3d:
        num_pts = 100
        xvals = np.linspace(xmin, xmax, num_pts)
        yvals = np.linspace(ymin, ymax, num_pts)
        (wxs, wys) = np.meshgrid(xvals, yvals)
        # w(1) * x1 ^ 2 + w(2) * x2 ^ 2 + w(3) * x1 + w(4) * x2 + w(5)
        w1 = lr.coef_[0][0]
        w2 = lr.coef_[0][1]
        w3 = lr.coef_[0][2]
        w4 = lr.coef_[0][3]
        w5 = lr.intercept_[0]
        wzs = np.array(map(expit, (wxs ** 2) * w1 + (wys ** 2) * w2 +
                           wxs * w3 + wys * w4 + w5))
        ax.plot_surface(wxs, wys, wzs)

    # plot predicted points
    if plot3d:
        ax.plot(xpred[:,0], xpred[:,1], ypred, color="lime", marker="o",
                ls="None")
    else:
        npred = len(xpred)
        for i in xrange(npred):
            if ypred[i] == 0:
                plt.plot(xpred[i][0], xpred[i][1], "bs")
            elif ypred[i] == 1:
                plt.plot(xpred[i][0], xpred[i][1], "rs")
            else:
                plt.plot(xpred[i][0], xpred[i][1], "ks")
    plt.show()

test4_circle()
